Encoder fix

mxl opened the pull request:
Related to but not fixes https://github.com/getquill/quill/issues/445

### Problem 1

I was looking for a solution to the above issue.

I have table in Postgresql database:

```sql
create table first_entity(strings text[] not null);
```

Because I can not define implicit encoder for `Traversable` that will take precedence over `io.getquill.context.async.Encoders.traversableEncoder` I decided to wrap `Traversable` type:

```scala
case class DbTraversable[T](v: Traversable[T])

trait Encoders extends io.getquill.context.async.Encoders {
  this: AsyncContext[_, _, _] =>
  
  implicit def dbTraversableEncoder[T](implicit e: Encoder[T]): Encoder[DbTraversable[T]] =
    encoder[DbTraversable[T]] { (value: DbTraversable[T]) =>
      value.v.map(_.asInstanceOf[Any]).toSeq
   }
}

trait Decoders extends io.getquill.context.async.Decoders {
  this: AsyncContext[_, _, _] =>

  implicit def dbTraversableDecoder[T]: Decoder[DbTraversable[T]] =
    new Decoder[DbTraversable[T]] {
      def apply(index: Int, row: RowData) = {
        row(index) match {
          case value: IndexedSeq[Any] =>
            DbTraversable[T](value.map(_.asInstanceOf[T]))
        }
      }
    }
}

object Main extends App {
  val context = new PostgresAsyncContext[SnakeCase]("database") with Encoders with Decoders
  import context._
  print(Await.result(
    context.run(quote(query[FirstEntity].insert))(
      List(FirstEntity(DbTraversable(Traversable("Hello", "world"))))).flatMap { _ =>
      context.run(quote(
        query[FirstEntity].map(e => e.strings)
      ))
    }, Duration.Inf))
}
```

It compiles but fails with an error:

>  com.github.mauricio.async.db.postgresql.exceptions.GenericDatabaseException: ErrorMessage(fields=Map(Detail -> Array value must start with "{" or di
> mension information., Line -> 289, File -> arrayfuncs.c, SQLSTATE -> 22P02, Routine -> array_in, Message -> malformed array literal: "DbTraversable(List(Hello, world))", 
> Severity -> ERROR))
> com.github.mauricio.async.db.postgresql.exceptions.GenericDatabaseException: ErrorMessage(fields=Map(Detail -> Array value must start with "{" or dimension information., 
> Line -> 289, File -> arrayfuncs.c, SQLSTATE -> 22P02, Routine -> array_in, Message -> malformed array literal: "DbTraversable(List(Hello, world))", Severity -> ERROR))
>         at com.github.mauricio.async.db.postgresql.PostgreSQLConnection.onError(PostgreSQLConnection.scala:172)

### Solution

I looked to `def encoder[T](f: T => Any): Encoder[T]` method and noticed that it does not apply `f`. I fixed that and it the error disappeared.

### Problem 2

Next I thought what if field is nullable:

```sql
create table first_entity(strings text[] null);
```

```scala
case class FirstEntity(strings: Option[DbTraversable[String]])

// Same Encoders and Decoders

object Main extends App {
  val context = new PostgresAsyncContext[SnakeCase]("database") with Encoders with Decoders
  import context._
  print(Await.result(
    context.run(quote(query[FirstEntity].insert))(
      List(FirstEntity(Some(DbTraversable(Traversable("Hello", "world")))))).flatMap { _ =>
      context.run(quote(
        query[FirstEntity].map(e => e.strings)
      ))
    }, Duration.Inf))
}
```

And it fails with the same error because `def optionEncoder[T](implicit d: Encoder[T])` does not apply `d` to `Option` value. I fixed that and the error disappeared.

### Checklist

- [x] Unit test all changes
- [ ] ~~Update `README.md` if applicable~~
- [ ] ~~Add `[WIP]` to the pull request title if it's work in progress~~
- [ ] ~~[Squash commits](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one) that aren't meaningful changes~~
- [x] Run `sbt scalariformFormat test:scalariformFormat` to make sure that the source files are formatted

@getquill/maintainers

fwbrasil commented:
@mxl thanks for the fix and especially the detailed description! :)

Could you add unit tests in these two files?

https://github.com/getquill/quill/blob/master/quill-async/src/test/scala/io/getquill/context/async/mysql/MysqlAsyncEncodingSpec.scala
https://github.com/getquill/quill/blob/master/quill-async/src/test/scala/io/getquill/context/async/postgres/PostgresAsyncEncodingSpec.scala

mxl commented:
@fwbrasil I added tests for encoding and decoding custom types. Also I fixed `Option` encoding for mirror and added customs types support for jdbc. I hope I done it well.

fwbrasil commented on quill-jdbc/src/main/scala/io/getquill/context/jdbc/JdbcEncoders.scala:
```diff
@@ -26,56 +30,44 @@ trait JdbcEncoders {
-            nullEncoder(idx, sqlType, row)
+          case None => d match {
+            case JdbcEncoder(sqlType) => nullEncoder(idx, sqlType, row)
+            case _                    => throw new NotImplementedError("")
```
Please use `fail` from `MessageUtil`. Also, could you provide a descriptive error message?

fwbrasil commented:
@mxl Thank you very much for the extra fixes! The build is failing because of unformatted files. Please run `sbt scalariformFormat test:scalariformFormat` again and push the branch.

mxl commented:
@fwbrasil Done. Also merged master.

fwbrasil commented:
👍
